cognite/fdm/general_domain/schema.py

```diff
@@ -58,37 +98,74 @@ def _register_type(cls: Type[DomainModelT]) -> Type[DomainModelT]:
         return _register_type
 
     def _do_register(self, name: str, cls: Type[DomainModelT]) -> None:
-        # Register type with strawberry schema
+        """Actually register a DomainModel with Strawberry."""
+
+        # this method could be called recursively, so avoid duplicate work:
         if name in self._processed_names:
             return
         self._processed_names.append(name)
 
+        # all field names on our DomainModel:
         field_names = [key for key in cls.__fields__ if key not in {"externalId"}]
-        # TODO handle custom types (JSON, Timestamp, etc.)
 
+        # find any custom scalar fields (e.g: Timestamp, JSONObject), they need special consideration:
+        scalar_fields = {}
+        for field_name in field_names:
+            field = cls.__fields__[field_name]
+            field_type = field.annotation
+            # find out if there is a custom scalar in this field's annotation:
+            while type_args := get_args(field_type):
+                field_type = type_args[0]
+            type_name = field_type.__name__
+            # if there is, make a corresponding strawberry scalar (only once), and
+            if type_name in scalars.__all__:
+                scalar_name = f"{type_name}Scalar"
+                if scalar_name not in self._scalar_names:
+                    self._scalar_names.append(scalar_name)
+                    globals().update({scalar_name: scalar(field_type)})
```

Comment:
# 👀 
Isn't there a long list of caveats when doing this?

Reply:
Well sure, but when you need it you need it :) But thinking about it, this is here only so that we dynamically create only those scalar classes that we need. It would be a minimal overhead to just import all of them and then this dynamic updating of `globals()` will not be needed (`import ...` will effectively do the same thing). Let's do that 👍

Reply:
I fully accept that sometimes hacky code is needed though 😄 ...but if you can make it more "vanilla" I would appreciate it! It also makes it less likely than non-CPython versions break 🚀

Reply:
Yeap, here it is: https://github.com/cognitedata/cognite-gql-pygen/pull/10/commits/b83e966b4382c4d9f514beb01e4370cf99070bdb